[Test] Introducing tier two functional test suite + several fixes

furszy opened the pull request:
Follow up to #1829, focused on decrease the testing time and continue improving the tier two functional tests.

Introduced the following changes:

1) Implemented a new RPC command `initmasternode` to initialize the masternode at runtime, with direct impact in the functional test time. The tests don't need to shut the nodes down and restart them anymore in order to setup the masternode/s.

2) Adapted `tiertwo_governance_sync_basic.py` to the new runtime initialization feature.

3) Further tier two cleanup. `Masternodeconfig` code refactored properly.

It's still in WIP (because it need some more work) but the results are very favorable for now:

Test previous time:
- 9:43 min

Test time with this changes:
- 3:21 min

Fuzzbawls reviewed (changes requested):
First pass, Concept ACK, but left comments where I think improvements and/or optimizations/clarifications could be made.

I also think that there may be no need to customize the regtest's `initialize_datadir` function as you can set multiple custom config params (per-node) already using the existing `self.extra_args` array-of-arrays.

random-zebra commented on src/rpc/masternode.cpp:
```diff
@@ -18,6 +18,22 @@
+    if (request.fHelp || (request.params.empty() || request.params.size() > 2)) {
+        throw std::runtime_error(
+                "initmasternode ( \"masternodePrivKey\" \"masternodeAddr\" )\n"
+                "\nInitialize masternode on runtime if it's not already initialized.\n"
```
I believe that "runtime" general meaning is "during the execution of the program" (as usually opposed to "compile time"), not just "during the program startup phase".
The most obvious example being `runtime_error`, which can be thrown at any time, not only during init.

If we want to be precise, runtime also includes the startup, so, even when reading the masternode.conf, technically, the masternodes are being initialized in the "runtime" phase.

Anyway, I think it's clear enough here, would just change it to either "**at** runtime" or "**on** demand".

furszy replied:
yeah, technically speaking, "at runtime" means essentially "while the program is running".
It covers the whole program lifecycle, beginning when is opened (or executed) and ending when is quit or closed.

I usually divide startup one-time arguments from the runtime definition as users cannot change them during the program execution. But that is purely a conceptual talk.
Every RPC command is obviously executed at runtime. So.. it's just redundant to say "at runtime" there. Agree on changing it to "on demand" so we can add a shutdown process in the same RPC command as well.

furszy commented:
just updated it with more changes. Decreased even more the test time from 4:22 to 3:21 minutes. Updated the PR description time as well.

This is still under WIP, will try to do the early feedback changes in the upcoming days 👍 . This is not priority to get merged anyway, Sapling PRs come first.

JSKitty commented on src/activemasternode.cpp:
```diff
@@ -14,6 +14,56 @@
+    CPubKey pubkey;
+
+    if (!CMessageSigner::GetKeysFromSecret(strMasterNodePrivKey, key, pubkey)) {
+        throw std::runtime_error(_("Invalid masternodeprivkey. Please see documenation."));
```
documenation --> documentation

JSKitty commented on src/rpc/masternode.cpp:
```diff
@@ -329,6 +346,17 @@ UniValue startmasternode (const JSONRPCRequest& request)
+                masternodeConfig.clear();
+                std::string error;
+                if (!masternodeConfig.read(error)) {
+                    throw std::runtime_error("Error realoding masternode.conf, " + error);
```
realoding --> reloading

JSKitty reviewed (approved):
Minor typo nits; everything else looks great!

furszy commented:
Updated per feedback plus added a basic budget finalization payment test coverage as well.

random-zebra commented on src/init.cpp:
```diff
@@ -1812,7 +1812,8 @@ bool AppInit2()
         LogPrintf("IS MASTER NODE\n");
-        initMasternode(GetArg("-masternodeprivkey", ""), GetArg("-masternodeaddr", ""), true);
+        auto res = initMasternode(GetArg("-masternodeprivkey", ""), GetArg("-masternodeaddr", ""), true);
+        if (!res) UIError(*res.error);
```
(07aec0cb4b1cd72f227b1d7626d7c69da0c6e9de)
Might be better to be sure that `res.error` is always initialized when`res.res` is false.
Easiest way is probably by just replacing the constructor `OperationResult(bool _res)` with a `OperationResult()`, that only sets `res` to `true`.
In other words, make it impossible to call `OperationResult(false)` (if the result is false, the error **must** be set).

random-zebra replied:
Or, even simpler... just make the optional error field private, and provide a guarded getter.
Added this in c3cf32326fb7d218ce3c1b2674e99ebf62076477.

random-zebra commented on src/rpc/budget.cpp:
```diff
@@ -712,14 +712,15 @@ UniValue mnfinalbudgetsuggest(const JSONRPCRequest& request)
 
-    budget.SubmitFinalBudget();
-    return NullUniValue;
+    uint256 budgetHash = budget.SubmitFinalBudget();
```
(4ff56e7beec709fb891ca54f83568b04662d60e5)
nit:
```suggestion
    const uint256& budgetHash = budget.SubmitFinalBudget();
```

random-zebra commented on .travis.yml:
```diff
@@ -158,6 +158,19 @@ jobs:
+        TEST_RUNNER_EXTRA="--tiertwo"  # Run tier two functional tests only.
+        GOAL="install"
+        BITCOIN_CONFIG="--with-gui=no --enable-glibc-back-compat --enable-reduce-exports --disable-online-rust --with-params-dir=$PARAMS_DIR"
+        BUILD_TIMEOUT=800
```
(f5fa73cdb04f05990c39c42c2426eb49668b37d8)
We don't have many tests yet, so for now we can probably raise this timeout (and not have to restart the build too often).

random-zebra reviewed (commented):
Looking good (modulo few nits).

Have some fixes and improvements for the python suite:
-  add an RPC to send the ping on demand
- abstract the functions in a new specific `PivxTier2TestFramework` based on mocked time
- refactor `masternode_activation` and `masternode_sync_basic` tests with the new framework
- add a separate test, `masternode_ping`, to only check the masternode ping thread real-time (with sleep)

Can either push them here, or open a new PR on top of this  one.

furszy commented:
awesome! 😄 go ahead!, either way is fine for me.

furszy reviewed (commented):
Overall the new commits are looking great 👌 , have done a quick pass through each of them. 
Have left few minor comments for us to check for when travis gets happy with this PR.

furszy commented:
PR title changed to reflect the current state. It got much bigger than what expected :) . Ready to go, we finally have the base framework and the first three functional tests for the tier two sub-systems :) .

Fuzzbawls commented on src/rpc/masternode.cpp:
```diff
@@ -18,6 +18,62 @@
+    }
+
+    if (!Params().IsRegTestNet()) {
+        throw JSONRPCError(RPC_MISC_ERROR, _("command available only for RegTest network"));
```
no need to translate this error message

Fuzzbawls commented on src/rpc/masternode.cpp:
```diff
@@ -18,6 +18,62 @@
+    }
+
+    if (!fMasterNode) {
+        throw JSONRPCError(RPC_MISC_ERROR, _("this is not a masternode"));
```
no need to translate this error message

Fuzzbawls commented on src/rpc/masternode.cpp:
```diff
@@ -923,9 +989,11 @@ static const CRPCCommand commands[] =
+    { "masternode",         "initmasternode",            &initmasternode,            true  },
 
-    /* Not shown in help */
+        /* Not shown in help */
```
nit: un-needed whitespace

Fuzzbawls commented on src/activemasternode.h:
```diff
@@ -18,13 +18,27 @@
 #define ACTIVE_MASTERNODE_NOT_CAPABLE 3
 #define ACTIVE_MASTERNODE_STARTED 4
 
+class OperationResult
```
#1798 also defines a class named `OperationResult` that conflicts with this.

Fuzzbawls commented on src/activemasternode.cpp:
```diff
@@ -14,6 +14,62 @@
 #include "netbase.h"
 #include "protocol.h"
 
+OperationResult errorOut(const std::string& errorStr)
```
#1798 also defines an `errorOut()` function in `sapling_operation.cpp` that conflicts with this.

Fuzzbawls reviewed (changes requested):
Overall looking good, but needs to be rebased on master to address duplicate class/function definitions. Also, left some minor nit comments.

furszy commented:
Updated per feedback, plus rebased it on master.

Note: added a commit dacfebe that is fixing a inv sync issue in master, introduced in the swiftx removal. I haven't pushed it in another PR because, as we have the new tier two test suite on this PR, it's fully verifiable here. Without it, the tier two sync simply doesn't work.

random-zebra reviewed (approved):
Some great catches near the end.
Tiertwo tests passing also with `-DDEBUG -DDEBUG_LOCKORDER`.
ACK 2b5fb9b2447611c6855b74d214a134b8922ec617
